rust-symcrypt/src/ecc/ecdsa.rs

```diff
@@ -0,0 +1,231 @@
+//! EcDsa functions. For further documentation please refer to symcrypt.h
+//!
+//! # Example
+//!
+//! ## ECDSA Sign and Verify
+//!
+//! ```rust
+//! use symcrypt::ecc::{EcKey, CurveType, EcKeyUsage};
+//! use hex::*;
+//!
+//! // Set up input hash value
+//! let hashed_message = hex::decode("4d55c99ef6bd54621662c3d110c3cb627c03d6311393b264ab97b90a4b15214a5593ba2510a53d63fb34be251facb697c973e11b665cb7920f1684b0031b4dd370cb927ca7168b0bf8ad285e05e9e31e34bc24024739fdc10b78586f29eff94412034e3b606ed850ec2c1900e8e68151fc4aee5adebb066eb6da4eaa5681378e").unwrap();
+//!
+//! // Generate a new ECDSA key pair
+//! let key = EcKey::generate_key_pair(CurveType::NistP256, EcKeyUsage::EcDsa).unwrap();
+//!
+//! // Sign the hash value
+//! let signature = key.ecdsa_sign(&hashed_message).unwrap();
+//!
+//! // Verify the signature
+//! let result = key.ecdsa_verify(&signature, &hashed_message);
+//!
+//! // Assert the signature is valid
+//! assert!(result.is_ok());
+//!
+//! ```
+//!
+use crate::ecc::{curve_to_num_format, CurveType, EcKey, EcKeyUsage};
+use crate::errors::SymCryptError;
+use std::vec;
+use symcrypt_sys;
+
+impl EcKey {
+    /// `ecdsa_sign()` returns a signature as a `Vec<u8>`, or a [`SymCryptError`] if the operation fails.
+    ///
+    /// `hashed_message` is a `&[u8]` that represents the hash value to sign.
+    ///
+    /// If the key usage is not [`EcKeyUsage::EcDsa`], or [`EcKeyUsage::EcDhAndEcDsa`] the function will
+    /// fail with a [`SymCryptError`] with the value [`SymCryptError::InvalidArgument`].
+    pub fn ecdsa_sign(&self, hash_value: &[u8]) -> Result<Vec<u8>, SymCryptError> {
+        // SymCrypt code AV's with SymCrypt Fatal Error.
+        // Panic'ing is not normal here in Rust so we are handling the error instead and returning SymCryptError::InvalidArgument
+        if self.get_ec_curve_usage() == EcKeyUsage::EcDh {
+            return Err(SymCryptError::InvalidArgument);
+        }
```

Comment:
This should not cause a fatal error.

```
    // Make sure that the key may be used in ECDSA
    if ( ((pKey->fAlgorithmInfo & SYMCRYPT_FLAG_ECKEY_ECDSA) == 0)  )
    {
        scError = SYMCRYPT_INVALID_ARGUMENT;
        goto cleanup;
    }
```

Reply:
Just tested this again with a debugger, in rust we panic here with `STATUS_STACK_BUFFER_OVERRUN`, will chat offline with you about this

Reply:
Weird. I'll file another SymCrypt bug on that. You can leave it as-is for now.